src/hd/usb.c

```diff
@@ -569,41 +569,48 @@ void add_input_dev(hd_data_t *hd_data, char *name)
       s
     );
 
-    for(hd = hd_data->hd; hd; hd = hd->next) {
-      if(
-        hd->module == hd_data->module &&
-        hd->sysfs_id &&
-        s &&
-        !strcmp(s, hd->sysfs_id)
-      ) {
-        t = NULL;
-        str_printf(&t, 0, "/dev/input/%s", sf_cdev_name);
+    /* find device (matching sysfs path) */
+    hd = hd_find_sysfs_id(hd_data, s);
+
+    /* not found? - retry one level up */
+    if(!hd) {
+      char *ns = new_str(s), *nt;
+      if((nt = strrchr(ns, '/'))) {
+        *nt = 0;
```

Comment:
I can't follow you; your implementation is way longer, not taking into account that there might be no '/' in the string, and much more obscure to me.

In my implementation it just duplicates a string and cuts off the end - perfectly clear.

`mem_free` works just like `free` - it doesn't care what's in the memory blob. Josef, you should do more C...